csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -2,195 +2,142 @@
                 {
-                    _ = StartedAsync(observer, activeServices);
+                    // We don't want to forward the cancel of a dispatch that has already completed.
+                    await observer.ServicesStartedAsync(services, cancel: default);
                 }
-
-                async Task StartedAsync(IServiceObserverPrx observer, string[] services)
+                catch (Exception ex)
                 {
-                    try
-                    {
-                        await observer.ServicesStartedAsync(services, cancel: cancel).ConfigureAwait(false);
-                    }
-                    catch (Exception ex)
-                    {
-                        RemoveObserver(observer, ex);
-                    }
+                    RemoveObserver(observer, ex);
                 }
             }
         }
 
-        public void Shutdown(Current current, CancellationToken cancel) => _ = _communicator.ShutdownAsync();
+        public ValueTask ShutdownAsync(Current current, CancellationToken cancel)
+        {
+            _ = _communicator.ShutdownAsync();
+            return default;
+        }
 
-        public void StartService(string name, Current current, CancellationToken cancel)
+        public async ValueTask StartServiceAsync(string name, Current current, CancellationToken cancel)
         {
-            ServiceInfo? info;
-            lock (_mutex)
+            // Search would be more efficient if services were contained in a map, but order is required for
+            // shutdown.
+            ServiceInfo? info = _services.Find(service => service.Name == name);
+            if (info == null)
             {
-                // Search would be more efficient if services were contained in a map, but order is required for
-                // shutdown.
-                info = _services.Find(service => service.Name.Equals(name));
-                if (info == null)
-                {
-                    throw new NoSuchServiceException();
-                }
-
-                if (info.Status != ServiceStatus.Stopped)
-                {
-                    throw new AlreadyStartedException();
-                }
-                info.Status = ServiceStatus.Starting;
-                _pendingStatusChanges = true;
+                throw new NoSuchServiceException();
             }
 
-            bool started = false;
             try
             {
-                Debug.Assert(info.Service != null);
-                Communicator? communicator = info.Communicator ?? _sharedCommunicator;
-                Debug.Assert(communicator != null);
-                info.Service.Start(info.Name, communicator, info.Args);
-                started = true;
+                await info.StartServiceAsync(_sharedCommunicator, cancel);
+
+                // We call ServiceStarted even if this call does not start the service - this is a way to refresh
+                // the status of this service in the observers.
```

Comment:
I don't find this comment very clear. Does this mean that `info.StartServiceAsync` might fail to start the service? But then isn't it supposed to throw?

Reply:
Ok I guess it's for the case where the service is already started in which case StartServiceAsync is a no-op and doesn't throw.

Reply:
Fixed comment